### PR TITLE
=cls #17846 Use CRDTs instead of PersistentActor to remember the state of the ShardCoordinator

### PR DESCRIPTION
**Pull request by oseval**
Currently this is not real PR. Only to show my attempt to change the claster-sharding dependency from the akka-persistence to the akka-ddata. I would incrementally update it when found time for this .

Idea:
1) Distributed state of the ShardRegions are all we need in the cluster-sharding.
2) Distributed state can be handled by akka-ddata module.
3) Inconsistency is possible if only we allowing two ShardRegions allocate equals Shards simultaneously.
4) It is possible avoid inconsistency if we have singleton Coordinator who can allow to the ShardRegion allocate the Shard. R1 ask C to allow him allocate the S1. C1 allow. R1 allocate S1 and then change DData + inner state. The Coordinator will be receiving these changes and then update inner state.
5) If the Coordinator fails, then it can be recovered from DData which be actual for all nodes inside a cluster. 

Welcome for comments!

It is very possible if I missing something in this concept. Correct me please.

And I am sorry for my English.

### CODE REVIEW

**Comment by smlin** [comment]
@patriknw I have some questions about cluster sharding. I need to know the "underlying idea". Currently only two questions.
1) Is it critical that the Coordinator reallocates the shards of the terminated regions immediately? Is it possible if the shard from a terminated region will be allocated only after sending a message to this shard?
2) Is it applicable if the terminated region would be remove itself from the distributed state by sending message to the replicator inside the `postStop`?

**Comment by patriknw** [comment]
Hi @smlin
I have not had time to look at your code yet. Will try to get here tomorrow.

There are two things that are persisted.
1. Coordinator state, the allocations of all shards
2. The started entities within each shard, see the Remember Entities section in the documentation.

Lets try to figure out the coordinator state first.
For that the allocation is always done for the first message to a shard, also for rebalance and crash.

It is possible to send message to replicator in postStop, but crash must also be handled and if the node is removed from the cluster the replicator might not be alive any more.

**Comment by smlin** [comment]
@patriknw PR is not ready for review yet. So you could do not waste your time. I will notify to you when it would be ready.

**Comment by patriknw** [comment]
@smlin My thinking around the design.

I think the coordinator should remain a singleton, because it is important with a central decision maker of where the shards are to be allocated. We could start with just saving the coordinator state in a LWWRegister, using WriteMajority and ReadMajority.

The LWWRegister is pretty safe here since the singleton failover is not instant, but we can later create a more customized data type that makes use of akka.cluster.Member.isOlderThan, we want to use the data from the newest singleton. That can be rather useful for singletons in general.

That has the small consistency problems that I sketched in the issue, but perhaps I'm looking too much for the bad scenarios, and perhaps we can make it more safe somehow.

The state of the `PersistenShard` could be stored in one `ORSet` per shard id.

I will be on vacation the next 4 weeks, so if I'm unresponsive here that is not because I'm not interested. Thanks for working on this.

**Comment by smlin** [comment]
@patriknw Thank you too.
I am was created first version of code. Currently it is mostly not completed and not works yet, but I wrote a lot of comments everywhere. And I was not working on PersistentShard because this is a much more work in a row - hope it could be doing step by step. I will be wait for response. 
Enjoy your holiday!
I will be working on this on my vacation at the next week:)

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 1]
```diff
@@ -387,112 +385,100 @@ class ShardCoordinator(typeName: String, settings: ClusterShardingSettings,
...
-          persistentState = persistentState.updated(evt)
-        case _: ShardHomeDeallocated ⇒
-          persistentState = persistentState.updated(evt)
+  override def receive: Receive = waitForState
+
+  def waitForState: Receive = {
+    case g @ GetSuccess(RegionsKey, _) =>
+      val regions: Map[ActorRef, Set[(ShardId, ActorRef)]] = g.get(RegionsKey).entries
+      regions.foreach { case (region, shards) =>
+        // TODO: Would the Coordinator notified by the `Terminated` message if the region was terminated before `watch`
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: yes, it will

**Comment by smlin** [comment]
I was done the first step as @patriknw  offered. I will be move the PersistentShard onto the CRDT in next step, since this is possible. 

Now PR is ready for review. 
@rkuhn Could somebody review this PR as long as Patrik on vacation?

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 2]
```diff
@@ -15,16 +15,24 @@ import akka.actor.Deploy
...
+import akka.cluster.ddata.Replicator.Get
+import akka.cluster.ddata.Replicator.GetSuccess
+import akka.cluster.ddata.Replicator.GetFailure
+import akka.cluster.ddata.Replicator.ModifyFailure
+import akka.cluster.ddata.Replicator.NotFound
+import akka.cluster.ddata.Replicator.ReadMajority
+import akka.cluster.ddata.Replicator.Update
+import akka.cluster.ddata.Replicator.UpdateSuccess
+import akka.cluster.ddata.Replicator.UpdateTimeout
+import akka.cluster.ddata.Replicator.WriteMajority
```
Comment (rkuhn) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: please use a wildcard import for these

File: project/AkkaBuild.scala [hunk 3]
```diff
@@ -136,9 +136,9 @@ object AkkaBuild extends Build {
     id = "akka-cluster-sharding",
     base = file("akka-cluster-sharding"),
     dependencies = Seq(cluster % "compile->compile;test->test;multi-jvm->multi-jvm",
-        persistence % "compile;test->provided", clusterTools)
+        persistence % "compile;test->provided", distributedData % "compile;test->provided", clusterTools)
```
Comment (rkuhn) [inline project/AkkaBuild.scala]: Is there still a dependency towards persistence?

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 4]
```diff
@@ -359,127 +368,116 @@ object ShardCoordinator {
...
+      state = g.get(CoordinatorStateKey).value
+      state.regionProxies.foreach(context.watch)
+      state.regions.foreach { case (a, _) ⇒ context.watch(a) }
+      state.shards.foreach { case (a, r) ⇒ sendHostShardMsg(a, r) }
       allocateShardHomes()
+      activate()
+
+    case GetFailure(CoordinatorStateKey, _) ⇒
+      log.error(
+        "The ShardCoordinator was unable to get an initial state within 'waiting-for-state-timeout' (was retried): {} millis",
```
Comment (rkuhn) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: “was retried” may be misunderstood, better would be “retrying”

**Comment by rkuhn** [comment]
Thanks a lot for this idea and PR, @smlin! There are some aspects that I cannot currently think through in full, revolving around the semantics of shard allocation. Currently you send back the result immediately before knowing that the CRDT update has been disseminated. The old implementation makes sure that the allocation has been persisted before it can be handed out and acted upon, which ensures that two different regions cannot create the same shard when the coordinator’s machine fails and a new coordinator makes a different decision.

The other point is related: the old implementation pulled shard location information lazily into the regions when needed, but CRDT replication would allow us to just listen to the same data at all regions, saving roundtrips to the coordinator.

One thing I really like about this approach is that the CRDT’s lifetime is coupled to the lifetime of the cluster itself. Persistence means that allocations can outlive the whole cluster, but when starting from scratch we may as well just reallocate things instead. The only missing feature then is that instantiated entities will not automatically be resurrected, but my thinking goes in the direction of making that an optional add-on since not everybody needs that and then we would have completely separated the sharding from the persistence concerns.

File: project/AkkaBuild.scala (hunk 3 above)
Comment (smlin) [inline project/AkkaBuild.scala]: Currently it is the "first step" and I removed the persistence just out of the coordinator. But the persistence stay alive inside the PersistenceShard. This is needed when `remember-entities = on`.

**Comment by smlin** [comment]
@rkuhn I have two ways. 
First - I have to fix the "double shards allocation" issue and write some test for checking this. And then I would be start the "next step" (and next PR) at bounds of which the distributed data could be wrote only inside their owners. The data about the allocated shards inside the shards, data about the started regions inside the regions. And of course I will to remove all dependencies onto the persistence.
Second - I can do all of this inside the current PR, but it is a lot of changes in a row. Of course I will do as you will say. But I think the first case is better.

**Comment by rkuhn** [comment]
Yes, I agree that the first proposal is the way to go, optimizing the use of the CRDTs can and should a second step; please create a ticket so that we have it properly on the radar.

File: akka-cluster-sharding/src/main/resources/reference.conf [hunk 5]
```diff
@@ -13,77 +13,83 @@ akka.cluster.sharding {
...
-    
+
     # The number of ongoing rebalancing processes is limited to this number.
     max-simultaneous-rebalance = 3
   }
-  
-  # Settings for the coordinator singleton. Same layout as akka.cluster.singleton.  
+
+  # Timeout of waiting the initial distributed state (an initial state will be queried again if the timeout happened)
+  waiting-for-state-timeout = 30 s
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/resources/reference.conf]: a more reasonable default is `5 s`

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 6]
```diff
@@ -359,127 +368,116 @@ object ShardCoordinator {
...
+        updatingStateTimeout.toMillis)
+      sendUpdate(evt)
+
+    case ModifyFailure(CoordinatorStateKey, error, cause, Some(evt: DomainEvent)) ⇒
+      log.error(
+        cause,
+        "The ShardCoordinator was unable to update a distributed state with error {}. Was retried with event {}.",
+        error,
+        evt)
+      sendUpdate(evt)
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: ModifyFailure happens only if **you** throw an exception from the the modify function, i.e. it will happen again if you try again. Better to stop or re-throw the cause

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 7]
```diff
@@ -359,127 +368,116 @@ object ShardCoordinator {
...
         gracefulShutdownInProgress -= region
-        saveSnapshotWhenNeeded()
-        persist(ShardRegionRegistered(region)) { evt ⇒
-          val firstRegion = persistentState.regions.isEmpty
+        val firstRegion = state.regions.isEmpty
 
-          persistentState = persistentState.updated(evt)
-          context.watch(region)
-          sender() ! RegisterAck(self)
+        update(ShardRegionRegistered(region))
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: Perhaps continue with watch and RegisterAck when you receive the `UpdateSuccess`? You can use the request (context) parameter in the `Update` message to pass along e.g. `sender`.

If you receive `UpdateTimeout` we should perhaps not retry the update here, since the region also has a retry mechanism.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 8]
```diff
@@ -359,127 +368,116 @@ object ShardCoordinator {
...
   var aliveRegions = Set.empty[ActorRef]
-  var persistCount = 0
 
   import context.dispatcher
   val rebalanceTask = context.system.scheduler.schedule(rebalanceInterval, rebalanceInterval, self, RebalanceTick)
 
   Cluster(context.system).subscribe(self, ClusterShuttingDown.getClass)
 
+  implicit val node = Cluster(context.system)
+  val CoordinatorStateKey = LWWRegisterKey[State](s"${typeName}CoordinatorState")
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: Later we could try to implement the member age aware register, using the `akka.cluster.Member.isOlderThan`, we want to use the data from the newest singleton. That can be rather useful for singletons in general.

If you want to take a stab at that new data type you find useful info here: http://doc.akka.io/docs/akka/snapshot/scala/distributed-data.html#Custom_Data_Type
and the following commit is a good example of what pieces that are required: 7041c76ba910e4474b1004060c5cbc08a121b5e1

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 9]
```diff
@@ -598,69 +587,74 @@ class ShardCoordinator(typeName: String, settings: ClusterShardingSettings,
...
+    state = state.updated(evt)
+    sendUpdate(evt)
+  }
+
+  def getState(): Unit =
+    replicator ! Get(CoordinatorStateKey, ReadMajority(waitingForStateTimeout))
+
+  def sendUpdate(evt: DomainEvent) =
+    replicator ! Update(CoordinatorStateKey, LWWRegister(State.empty), WriteMajority(updatingStateTimeout), Some(evt)) { reg ⇒
+      reg.withValue(reg.value.updated(evt))
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: shouldn't this be `reg.withValue(state)`?

I think the approach of applying the event again is problematic when you retry the updates. ordering, duplicate operations?

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 10]
```diff
@@ -359,127 +368,116 @@ object ShardCoordinator {
...
+      activate()
+
+    case GetFailure(CoordinatorStateKey, _) ⇒
+      log.error(
+        "The ShardCoordinator was unable to get an initial state within 'waiting-for-state-timeout' (was retried): {} millis",
+        waitingForStateTimeout.toMillis)
+      getState()
+
+    case NotFound(CoordinatorStateKey, _) ⇒ activate()
+    case m                                ⇒ stash()
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: Stashing here is a bit dangerous since it might take a long time until we receive a successful Get(majority), e.g. when nodes are unreachable (but not yet removed from cluster). Can't we just drop all other messages? They are resent anyway, right?

**Comment by patriknw** [comment]
@smlin I took a first look. I think it is in the right direction. I appreciate the step-wise approach you take on this. Great work!

To set the expectations right, we will probably have to support both this and the persistent sharding and we need to discuss how to package those. Anyway, don't worry about that now. Please continue in this PR, and separate the different steps with separate commits.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala (hunk 9 above)
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: I mean:

    val s = state
    replicator ! Update(...) { reg ⇒
      reg.withValue(s)
    }

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala (hunk 6 above)
Comment (smlin) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: Not exactly so. It depends on state inside - if after the failure there was another successful update then at next time the failure could be not occurs.
Anyway I think errors in logs slightly better then kill cluster-sharding at all.

**Comment by smlin** [comment]
@rkuhn @patriknw I am stopped on consistency requirements to case of duplicated shards. To avoid such duplication a coordinator must ensure changes of state to a whole cluster (WriteAll in case of CRDTs) or synchronize state with all regions before moving to the active state. But in second step I will change the source of distributed state changes from a coordinator to the regions and shards. So after second step duplication will be avoided automatically "by design". 
Now question: is WriteMajority enough for merge this PR or it must be implemented with the guaranteed consistency?
Currently I think that this step is finished and need for review before start the second step.

**Comment by smlin** [comment]
I have to notice one case when avoiding the shards duplication is impossible. If cluster was fragmented (due to broken link between two datacenters or some other cases), then such situation would be end up with the two identical coordinators in the both clusters. 
In case when coordinator extends the PersistentActor this will lead to crash both coordinators (due concurrent writes by two persistent actors) and could be repaired only by cleaning persistent data from database. 
In case of CRDT this eventually would lead to stand up two identical shards/entities and could be crash these entities if they are persistent actors.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 11]
```diff
@@ -359,127 +359,129 @@ object ShardCoordinator {
...
+          sendUpdate(evt)
+      }
+
+    case ModifyFailure(CoordinatorStateKey, error, cause, Some(evt: DomainEvent)) ⇒
+      log.error(
+        cause,
+        "The ShardCoordinator was unable to update a distributed state with error {}. Was retrying with event {}.",
+        error,
+        evt)
+      sendUpdate(evt)
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: `ModifyFailure` only happens if you modify function in the `Update` message throws an exception, and that should never happen unless you have done a programming error. No point in trying to do it again. Will probably fail again. Instead you should just re-throw the cause.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 12]
```diff
@@ -359,127 +359,129 @@ object ShardCoordinator {
...
-        case ShardRegionProxyTerminated(proxy) ⇒
-          if (persistentState.regionProxies.contains(proxy))
-            persistentState = persistentState.updated(evt)
-        case ShardHomeAllocated(shard, region) ⇒
-          persistentState = persistentState.updated(evt)
-        case _: ShardHomeDeallocated ⇒
-          persistentState = persistentState.updated(evt)
+          val firstRegion = state.regions.isEmpty
+
+          state = state.updated(evt)
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: The state was already updated in the [update](https://github.com/akka/akka/pull/17871/files#diff-e3b09e26a65bebe3b08cd0b802fde5e8R601) method.

The order of the events applied to the state is important. The original order can be changed if we apply the event here, because of the [retry](https://github.com/akka/akka/pull/17871/files#diff-e3b09e26a65bebe3b08cd0b802fde5e8R445) and that UpdateSuccess messages are not guaranteed to be received in the same order as corresponding Update messages were emitted.

I see two options.
1) Apply the even and set the local state before sending Update to the replicator. Some of the code here must also be moved to that point. No retry. We rely on the gossip replication in case WriteMajority timeout.
2) We apply the event and set the local state here instead of doing it before sending Update to the replicator. We retry until UpdateSuccess. We stash all incoming messages until we receive UpdateSuccess, i.e. we perform only one update at a time.

The second alternative is more safe in case of fail over to another coordinator instance.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 13]
```diff
@@ -359,127 +359,129 @@ object ShardCoordinator {
...
+      activate()
+
+    case GetFailure(CoordinatorStateKey, _) ⇒
+      log.error(
+        "The ShardCoordinator was unable to get an initial state within 'waiting-for-state-timeout' (was retrying): {} millis",
+        waitingForStateTimeout.toMillis)
+      getState()
+
+    case NotFound(CoordinatorStateKey, _) ⇒ activate()
+  }
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: perhaps a comment that it is alright to drop all other messages, they will be retried

**Comment by patriknw** [comment]
> then such situation would be end up with the two identical coordinators in the both clusters

We must assume that cluster singleton keeps it promise to be a singleton. There is not much we can do if  the cluster is split into two separate clusters each hosting its own singleton instance. That will not happen by the network partition itself, but by the used downing strategy. That is why we recommend against using auto-downing when using Akka Persistence. We are attacking that issue from other angles.

We never merge half complete solutions, so please continue to update this PR with more commits for the additional steps.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala (hunk 11 above)
Comment (smlin) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: Could it happens due to message/updates reordering? And would fixed after some time.

**Comment by smlin** [comment]
> We are attacking that issue from other angles.
Am I right understand this mean that issue must not be solved in this PR?

**Comment by patriknw** [comment]
> Am I right understand this mean that issue must not be solved in this PR?

yes

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala (hunk 11 above)
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: I don't think so, typical error would be things like `NullPointerException`, i.e. it should never happen unless we have a bug somewhere and then we don't want to continue anyway

**Comment by smlin** [comment]
@patriknw It turns out that ddata realization (without straightforward message exchanges between coordinator, regions and shards) is tremendously slow since gossips have a timeout. So I see two variants:

1) The `Replicator` must sending gossips immediately to the nodes where subscribers exists.
2) Currently this PR seems to be working and I think it is anyway better then saving a state to disk. 
Using a ddata not only for the coordinator state gives slightly more safety and no more.

Probability of the shards duplication even at the moments of cluster failures seems to be very small. I don't know exactly, but it possible that such failure could happen only if `auto-down` enabled. But if it is enabled then saving the coordinator state to disk also have the problems.

Currently I think it is complete solution. Any better solution required absolutely different idea and realization. Sorry, I have no a time to do such realization.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala [hunk 14]
```diff
@@ -155,6 +155,12 @@ object ShardRegion {
   @SerialVersionUID(1L) final case object GracefulShutdown extends ShardRegionCommand
 
   /**
+   * We must be sure that a shard is initialized before to start send messages to it.
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala]: what is the reason for the changes in this file?

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala [hunk 15]
```diff
@@ -96,7 +96,7 @@ private[akka] class Shard(
   extractShardId: ShardRegion.ExtractShardId,
   handOffStopMessage: Any) extends Actor with ActorLogging {
 
-  import ShardRegion.{ handOffStopperProps, EntityId, Msg, Passivate }
+  import ShardRegion.{ handOffStopperProps, EntityId, Msg, Passivate, ShardInitialized }
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala]: what is the reason for the changes in this file?

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 16]
```diff
@@ -359,127 +359,117 @@ object ShardCoordinator {
  * @see [[ClusterSharding$ ClusterSharding extension]]
  */
 class ShardCoordinator(typeName: String, settings: ClusterShardingSettings,
-                       allocationStrategy: ShardCoordinator.ShardAllocationStrategy)
-  extends PersistentActor with ActorLogging {
+                       allocationStrategy: ShardCoordinator.ShardAllocationStrategy,
+                       replicator: ActorRef)
+  extends Actor with ActorLogging with Stash {
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: In 2.4 we need to support both the old persistent implementation and the new ddata implementation. The reason is that akka-cluster-sharding is not experimental and therefore it cannot by default use the experimental distributed-data module.

Would it be possible to extract the specific persistent/ddata parts to separate subclasses? AbstractShardCoordinator, PersistentShardCoordinator, DDataShardCoordinator.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 17]
```diff
@@ -598,69 +579,76 @@ class ShardCoordinator(typeName: String, settings: ClusterShardingSettings,
     case _: CurrentClusterState ⇒
   }
 
+  def activate() = {
+    context.become(active)
+    unstashAll()
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: is stash used at all?

**Comment by patriknw** [comment]
@smlin I agree that the solution in this PR is good and it is a great advantage that it is so similar to current persistent implementation, since we need to support both approaches in 2.4 (as described in my comment above). Would you be able to wrap up that part or would you like that I make those adjustments?

**Comment by smlin** [comment]
I will try do it by myslef. Thank you for review.

**Comment by patriknw** [comment]
Thanks!

**Comment by patriknw** [comment]
To clarify my idea. We should make it possible to switch between the two implementations with a config setting. The dependency to distributed-data should be in scope provided, so the user must define the dependency and switch the config setting to use this implementation. Does that make sense?

**Comment by smlin** [comment]
Yes, it make sense. Perhaps now it could be done also for akka-persistence dependency.

**Comment by smlin** [comment]
@patriknw I wrote two implementations of the coordinator which depends on config parameter. I duplicated tests for ddata coordinator - currently I don't know how make it more elegantly. Possible you could help me.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala (hunk 15 above)
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala]: I would still like to understand why this was changed

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala [hunk 18]
```diff
@@ -393,62 +396,19 @@ class ShardCoordinator(typeName: String, settings: ClusterShardingSettings,
...
-        persist(ShardRegionRegistered(region)) { evt ⇒
-          val firstRegion = persistentState.regions.isEmpty
-
-          persistentState = persistentState.updated(evt)
+        update(ShardRegionRegistered(region)) { evt ⇒
+          val firstRegion = state.regions.isEmpty
+          state = state.updated(evt)
           context.watch(region)
-          sender() ! RegisterAck(self)
+          region ! RegisterAck(self)
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala]: sender() and region is the same, right?
but now we use region here, but sender() a few lines above

**Comment by patriknw** [comment]
@smlin excellent separation! I have a few comments, but it looks like this can be merged soon. We are releasing on 2.4-RC1 on Friday so it would be great if we can merge it on Thursday at the latests.

Could you also add a section in the docs of the alternative ddata mode:
* akka-docs/rst/scala/cluster-sharding.rst
* akka-docs/rst/java/cluster-sharding.rst

Regarding the tests I think you could split them up into an abstract class and two implementations. The only difference should be the config (and some initial creation of actors). See example of similar split:

* https://github.com/akka/akka/blob/master/akka-persistence/src/test/scala/akka/persistence/PersistentActorSpec.scala#L1139
* https://github.com/akka/akka/blob/master/akka-cluster/src/multi-jvm/scala/akka/cluster/ClientDowningNodeThatIsUnreachableSpec.scala#L22-L34

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala (hunk 15 above)
Comment (smlin) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala]: Sorry, I forgot answer to you here.
Sometimes persistent shard could be failure during a replay process due to temporary journal unavailability. The user messages which sent to such shard just after creation, but before a failure would be lost. This code removes that problem.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala [hunk 19]
```diff
@@ -302,11 +308,13 @@ class ShardRegion(
 
   def receive = {
     case Terminated(ref)                         ⇒ receiveTerminated(ref)
+    case ShardInitialized(shardId)               ⇒ initializeShard(shardId, sender())
     case evt: ClusterDomainEvent                 ⇒ receiveClusterEvent(evt)
     case state: CurrentClusterState              ⇒ receiveClusterState(state)
     case msg: CoordinatorMessage                 ⇒ receiveCoordinatorMessage(msg)
     case cmd: ShardRegionCommand                 ⇒ receiveCommand(cmd)
     case msg if extractEntityId.isDefinedAt(msg) ⇒ deliverMessage(msg, sender())
+    case msg: RestartShard                       ⇒ deliverMessage(msg, sender())
```
Comment (smlin) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala]: The `RestartShard` message handling has not existed before - it is a bug fix.

Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala]: great catch, it's scheduled by `receiveTerminated`

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala (hunk 15 above)
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala]: good catch
note for other reviewers: this is handled in ShardRegion.scala

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala (hunk 14 above)
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala]: as mentioned elsewhere: Sometimes persistent shard could be failure during a replay process due to temporary journal unavailability. The user messages which sent to such shard just after creation, but before a failure would be lost. This code removes that problem.

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala [hunk 20]
```diff
@@ -560,9 +590,10 @@ class ShardRegion(
...
           name))
-        shards = shards.updated(id, shard)
         shardsByRef = shardsByRef.updated(shard, id)
-        shard
+        None
+      case Some(props) ⇒
+        None
       case None ⇒
         throw new IllegalStateException("Shard must not be allocated to a proxy only ShardRegion")
     })
```
Comment (patriknw) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala]: The changes in this file looks correct

**Comment by patriknw** [comment]
LGTM
@smlin please squash into one commit and I will merge it. Excellent contribution!

I will add a few notes about experimental and such in the docs, but I can take care of that.

**Comment by smlin** [comment]
@patriknw Squashed.

**Comment by patriknw** [comment]
thanks!

**Comment by patriknw** [comment]
Refs #17846
